### ISSUE TITLE
Distributed(c10d) occasionally locks up for mixed precision training

### ISSUE THREAD

**Issue by bearpelican** [pytorch/pytorch#11672]
After creating DistributedDataParallel model, pytorch freezes when trying to create float tensors.
This happens 50% of the time.
Freeze happens on this [line](https://gist.github.com/bearpelican/4adcadce7f0687681bc9be2324ebafc8#file-train_c10d_deadlock-py-L73)

Strangely, it only happens when I set batch norm layers to full precision. Uncommenting out this [line](https://gist.github.com/bearpelican/4adcadce7f0687681bc9be2324ebafc8#file-train_c10d_deadlock-py-L66
) works fine.

Hardware - p3.16xlarge (8 V100 GPUs)
Version: Pytorch 0.5.0a0+57f149a
NCCL 2.2.13, Cuda 9.2, Cudnn 7.2.1

Distributed command run:
NCCL_DEBUG=VERSION python -m torch.distributed.launch --nproc_per_node=8 --nnodes=1 --node_rank=0 --master_addr=[machineip] --master_port=[port] train_minimal.py --dist-url env://

**Comment by bearpelican** [comment]
@teng-li

**Comment by bearpelican** [comment]
Stack trace:
  ```File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/site-packages/torch/distributed/launch.py", line 211, in <module>
    main()
  File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/site-packages/torch/distributed/launch.py", line 207, in main
    process.wait()
  File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/subprocess.py", line 971, in wait
    return self._wait(timeout=timeout)
  File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/subprocess.py", line 1601, in _wait
    (pid, sts) = self._try_wait(0)
  File "/home/ubuntu/anaconda3/envs/pytorch_source/lib/python3.7/subprocess.py", line 1559, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)```

**Comment by bearpelican** [comment]
Yup Let me try that

**Comment by yaroslavvb** [comment]
I think that's just the stack trace of the launcher process that's waiting for children python processes. To figure out why child python process is hanging you could get its pid and do 

gdb -p <pid>
# for main thread
gdb bt
# or for all threads
gdb apply all bt

**Comment by pietern** [comment]
@bearpelican @yaroslavvb These being full on processes I don't think they will show up in gdb for the parent. You can use ps to figure out the PIDs of the child processes and run gdb on them separate to figure out where it is hanging.

**Comment by bearpelican** [comment]
Setting ```mode = CUDNN_BATCHNORM_SPATIAL``` seems to fix it for my minimal example on a single machine:
https://gist.github.com/bearpelican/4adcadce7f0687681bc9be2324ebafc8#file-train_c10d_deadlock-py-L73

However, when I move to 4 machines, each machine still hangs. 
Not sure why it would make a difference. I'll run gdb on the child threads to see

**Comment by bearpelican** [comment]
Looks to be fixed now!

**Comment by zimenglan-sysu-512** [comment]
hi @bearpelican 
how did u solve the problem?

**Comment by yaroslavvb** [comment]
It went away after upgrading

**Comment by pietern** [comment]
@zimenglan-sysu-512 @yaroslavvb This was likely fixed in #13496.

**Comment by teng-li** [comment]
Yes, this is the fix: https://github.com/pytorch/pytorch/pull/13496

**Comment by bearpelican** [comment]
Hmm looks like made a mistake and didn't fully retest this. Sorry about that!

C10D hangs for multi-gpu + multi-machine. As before, it only happens when batch-norm is set to full precision. Note: it works fine for single-machine, multi-gpu now (as #13496 seems to fix that).

Deadlock happens on this line:
https://gist.github.com/bearpelican/c93ebe5513417055582e6137eedd514f#file-pytorch_minimal_4machine-py-L85

I believe this is the stacktrace:
https://gist.github.com/bearpelican/4808a4d0b3008f5afc068104e5876475

**Comment by pietern** [comment]
The stack trace shows the copy is blocked on something, so the actual problem will be elsewhere. Can you try and run after setting the environment variable `CUDA_LAUNCH_BLOCKING=1`? Perhaps this will uncover the underlying call that's the source of this problem...

**Comment by ngimel** [comment]
Looks like broadcast in DistributedDataParallel initialization never finishes.

**Comment by teng-li** [comment]
Can you get the NCCL log
try setting both NCCL_DEBUG=INFO and NCCL_DEBUG_SUBSYS=COLL

**Comment by bearpelican** [comment]
Yup! Here's the log with environment variables set:
https://gist.github.com/bearpelican/eb2bc2bc1818581efef91172277e4b3b

**Comment by bearpelican** [comment]
If you have access to aws and p3 instance capacity, I just created a simple breaking example using @yaroslavvb 's distributed framework 
https://github.com/bearpelican/distributed_minimal

**Comment by teng-li** [comment]
@bearpelican Ok, let me try to repro on my own machine.

**Comment by teng-li** [comment]
@bearpelican  I am not able to repro on my end, and the code works fine on two DGX-1s.  The NCCL log doesn't output anything that is not normal.

**Comment by teng-li** [comment]
What GPUs are these on AWS?

**Comment by bearpelican** [comment]
Shoot, looking at my version hash, looks like I may not have update pytorch correctly. 
I'll rebuild and install.

Good chance the issue is fixed already, as you said.

**Comment by bearpelican** [comment]
Yeah looks like my conda paths got messed up. Apologize for the false alarm 🙇
